### ISSUE TITLE
Confusing print out of parent package enables when only subpackages are enabled

### ISSUE THREAD

**Issue by ibaned** [TriBITSPub/TriBITS#188]
my Trilinos configuration is currently giving output like this:
```
Enabling all optional intra-package enables <TRIBITS_PACKAGE>_ENABLE_<DEPPACKAGE> that are not currently disabled if both sets of packages are enabled ...

-- Setting TeuchosCore_ENABLE_KokkosCore=ON since Trilinos_ENABLE_TeuchosCore=ON AND Trilinos_ENABLE_KokkosCore=ON
-- Setting Sacado_ENABLE_KokkosCore=ON since Trilinos_ENABLE_Sacado=ON AND Trilinos_ENABLE_KokkosCore=ON
-- Setting Sacado_ENABLE_TeuchosCore=ON since Trilinos_ENABLE_Sacado=ON AND Trilinos_ENABLE_TeuchosCore=ON

...

Enabling all parent packages that have at least one subpackage enabled ...

-- Setting Trilinos_ENABLE_Kokkos=ON because Trilinos_ENABLE_KokkosCore=ON
-- Setting Trilinos_ENABLE_Teuchos=ON because Trilinos_ENABLE_TeuchosCore=ON
```

Sacado declares an optional dependency on parent package Teuchos, but `Sacado_ENABLE_Teuchos` is *not* to set to  `ON` despite the fact that parent package enabling does set `Trilinos_ENABLE_Teuchos=ON`.  This seems like wrong behavior from TriBITS. If `Teuchos` and `Sacado` both end up in the final set of enabled packages, then `Sacado_ENABLE_Teuchos` should definitely be set to `ON`.

This causes trouble later down the line because Sacado relies heavily on `HAVE_SACADO_TEUCHOS` to `#ifdef` in important code, and no amount of work in `Dependencies.cmake` can activate this right now; I have to explicitly set `Trilinos_ENABLE_Teuchos` in my `do-config.sh`.

**Comment by bartlettroscoe** [comment]
@ibaned, can you please attach the full configure output?  From that I can explain what is happening here and how to get the behavior you want.

Before going into the justification for this behavior, to fix this, Sacado needs to explicitly list what Teuchos subpackages it depends on and then it needs to use conditional logic on those SE packages, not the parent package.  For example, if Sacado depends on `TeuchosCore`, then it needs to use `Sacado_ENABLE_TeuchosCore` in internal logic and **NOT** `Sacado_ENABLE_Teuchos` if that is in fact the set of software it depends on.  It makes no sense to declare a dependency on an outer parent package unless you actually depend on **all** of the required subpackages of that parent package (and I hope that is not the case for the Sacado dependency on Teuchos).

In general, all Trilinos packages need to be changed to depend on subpackages and never depend on the parent packages.   For example, given that Teuchos is broken down into subpackages, no downstream Trilinos package should every depend on 'Teuchos'.  Instead, they should explicitly list the Teuchos subpackages they depend on like `TeuchosCore` or `TeuchosComm`.   Subpackages are designed for dependency management.  Parent packages are supposed to be a means to aggregate related subpackages, help organize them in enable and disable logic (e.g. allow you to enable or disable all of them at once), and be a point of documentation for the related subpackages.

Now for an explanation of this behavior and suggestions for how to improve this ...

So this is not really a defect in behavior.  This is the (I would argue) correct behavior.  This is, perhaps, a defect in terminology and the clarity of the output.  This behavior is called out and explained (a little) in:

* https://tribits.org/doc/TribitsDevelopersGuide.html#subpackage-enable-does-not-auto-enable-the-parent-package

The problem is how to handle and represent the relationship between subpackages and parent packages and dependencies management.  The problem is the interpretation of the output:

```
Enabling all parent packages that have at least one subpackage enabled ...

-- Setting Trilinos_ENABLE_Kokkos=ON because Trilinos_ENABLE_KokkosCore=ON
-- Setting Trilinos_ENABLE_Teuchos=ON because Trilinos_ENABLE_TeuchosCore=ON
```

What this really means is that the parent packages 'Kokkos' and 'Teuchos' will get processed, but only for the subpackages that were enabled in the above traced logic.  (The other subpackages that were not enabled, will not get processed.)  This is done so that 'Kokkos' and 'Teuchos' are shown as enabled parent packages in the line "Final set of enabled packages" (because they have 1 or more of their subpackages enabled).   But you will see that only a subset of the 'Teuchos' subpackages actually get processed.  Otherwise, you see 'TeuchosCore' listed in "Final set of enabled SE packages" but you don't see 'Teuchos' listed in "Final set of enabled packages".  That is how TriBITS used to show things years ago but I found that more confusing.

The question is how can TriBITS output be changed to make this behavior more obvious?  I am open to suggestions.

**Comment by ibaned** [comment]
I've attached my configure outpur, including the explicit Teuchos enable. I was trying to do MiniTensor development, and MiniTensor tests rely on Sacado functionality which is only available when `HAVE_SACADO_TEUCHOS` is defined. That is a separate issue.
[config_log.txt](https://github.com/TriBITSPub/TriBITS/files/955933/config_log.txt)

Okay, so there is a fundamental ambiguity between a parent package representing *all* of its subpackages or *any* of its subpackages. `Trilinos_ENABLE_Teuchos` will be `ON` if *any* of Teuchos is enable, while `Sacado_ENABLE_Teuchos` will only be `ON` if *all* of Teuchos is enabled. They look so similar, there is no intuitive way to guess that that is their behavior.

There are two things I think can be improved without changing the behavior:
1. CMake warning when parent packages are listed in `Dependencies.cmake` instead of SE packages. Something like:
```
Sacado: optional library dependence Teuchos is a parent package, it is recommended to list SE packages only.
```
2. configure-time output like this:
```
-- NOT setting Sacado_ENABLE_Teuchos=ON because TeuchosNumerics is not enabled
```

In short, the output should (1) discourage the dependence on a parent package and (2) be clear that the dependence on a parent package will only be satisfied when all its sub-packages are enabled.

By the way, if I manually enable all the Teuchos sub-packages, should that actually trigger `Sacado_ENABLE_Teuchos=ON` ?

**Comment by bartlettroscoe** [comment]
@ibaned, these are good suggestions.  Some specific responses/questions ...

> Okay, so there is a fundamental ambiguity between a parent package representing all of its subpackages or any of its subpackages

Yup, that is the problem.   But if we only use subpackages for defining dependencies between SE packages, then the ambiguity largely goes away.  But we had to allow downstream packages to list upstream parent packages in order to make breaking a package into subpackages a backward compatible operation.  That need for smooth backward compatibility when breaking a package into subpackages is the reason for the current (perhaps confusing) behavior. 

> By the way, if I manually enable all the Teuchos sub-packages, should that actually trigger `Sacado_ENABLE_Teuchos=ON` ?

No, it will not.  What the TriBITS dependency system states in:

* https://tribits.org/doc/TribitsDevelopersGuide.html#enable-disable-of-parent-package-is-enable-disable-for-subpackages

is "Enable/disable of parent package is enable/disable for subpackages".  It does **NOT** state the converse "Enable of all subpackages is equivalent to enable for the parent package".  I think it might be confusing (and a bit magical) if it did.  For example, if a new subpackage got added to Teuchos, then Teuchos would go from being enabled to no longer being enabled when the other existing subpackages were enabled..

Given that enabling all of a parent package's subpackages is not equivalent to enabling the parent package in all contexts, when you give the example:

```
-- NOT setting Sacado_ENABLE_Teuchos=ON because TeuchosNumerics is not enabled
```

should it not be something like:

```
-- NOTE: NOT setting Sacado_ENABLE_Teuchos=ON because Teuchos is not enabled at this point!
```

?

Then we would perhaps change the output:

```
Enabling all parent packages that have at least one subpackage enabled ...

-- Setting Trilinos_ENABLE_Kokkos=ON because Trilinos_ENABLE_KokkosCore=ON
-- Setting Trilinos_ENABLE_Teuchos=ON because Trilinos_ENABLE_TeuchosCore=ON
```

to something like:

```
Enabling shell of parent packages (mostly for show) that have at least one subpackage enabled ...

-- Setting Trilinos_ENABLE_Kokkos=ON because Trilinos_ENABLE_KokkosCore=ON
-- Setting Trilinos_ENABLE_Teuchos=ON because Trilinos_ENABLE_TeuchosCore=ON
```

Hopefully that would be somewhat less confusing.

As for:

> CMake warning when parent packages are listed in Dependencies.cmake instead of SE packages. Something like:
>
> ```
> Sacado: optional library dependence Teuchos is a parent package, it is recommended to list SE package
> ```

I think that type of warning might be useful in development mode.  But since it would be executed no matter what packages got enabled, it would be very verbose for Trilinos currently.  The problem is that similar warnings have gone ignored and then they just clutter up the configure output and people complain.  For example, we had that issue with missing SCOREC packages (so we made those NOTES go away).  And we still see a similar problem with improper usage of DEPLIBS for the TRIBITS_ADD_LIBRARY() and TRIBITS_ADD_EXECUTABLE() commands in several packages and warnings that TriBITS generates for these. But if we add this NOTE in development-mode, then someone will need to update all of Trilinos to depend on subpackages instead of parent packages.  Otherwise the warnings will just clutter up the output and people will get mad (and nothing will get better).  Therefore, I am not sure we should add this output until we can convince Trilinos developers to clean up their package dependencies (and make them more narrow and target).  Also, would this problem hopefully be hinted at by seeing output like "NOTE: NOT setting Sacado_ENABLE_Teuchos=ON because Teuchos is not enabled at this point!"?

So, a compromise might be:

1) Print output like `NOTE: NOT setting Sacado_ENABLE_Teuchos=ON because Teuchos is not enabled at this point!` to help developers and users understand why that enable is not occurring (and should hopefully give the downstream package developers a hint that something should be changed in their Dependencies.cmake files)

2) Change the output "Enabling all parent packages that have at least one subpackage enabled ..." to "Enabling shell of parent packages (mostly for show) that have at least one subpackage enabled ...".

Does that sound like a reasonable compromise?  If so, I think this can be implemented pretty quickly and yet would not create too much extra output that might make people mad.

As for this particular issue with Sacado, MiniTensor and Teuchos, it it clear how to address the current problem you are having?  If not, let's create a Trilinos GitHub issue to address that specific issue there.

**Comment by ibaned** [comment]
Thanks @bartlettroscoe . Item (1) of your compromise is what is most important to me, that way I would have seen the cause of my troubles. Item (2) is good, especially because doing item (1) without doing (2) will make the output really confusing.

CMake warnings are just one possible method for getting people to change how they do things, if it hasn't worked historically then lets use another method. I've opened trilinos/Trilinos#1263 to ask Sacado to change their dependencies to be in terms of sub-packages. Until they do that, users must simply remember to explicitly enable the Teuchos parent package when compiling MiniTensor with tests enabled.

**Comment by bartlettroscoe** [comment]
>  Item (1) of your compromise is what is most important to me, that way I would have seen the cause of my troubles. Item (2) is good, especially because doing item (1) without doing (2) will make the output really confusing.

Okay, I will make those changes based on the plan in this issue.  Should not take too long.

> CMake warnings are just one possible method for getting people to change how they do things, if it hasn't worked historically then lets use another method.

GitHub issues only work if people respond to them.  Lots of examples where they get ignored even more in Trilinos.

But I can see that for helping to debug depending on parent packages instead of correctly depending on subpackages (which is why subpackages were invented in the first place), it would be helpful for TriBITS to print out these warnings automatically.  Therefore, I will see about adding that option (e.g. `<Project>_WARN_ABOUT_DIRECT_PARENT_PACKAGE_DEPENDENCIES`).

> I've opened trilinos/Trilinos#1263 to ask Sacado to change their dependencies to be in terms of sub-packages.  Until they do that, users must simply remember to explicitly enable the Teuchos parent package when compiling MiniTensor with tests enabled.

We might help that along with a PR.  If you don't have time for that, I could take a quick look to see how long it would take to fix (which is likely not long).  I just did some analysis on that and it seems pretty straightforward.

**Comment by bartlettroscoe** [comment]
I just pushed the TriBITS commit 679fb22227cb3e68b8443b2cd8307f8187031cde that addresses this.  Once I snapshot TriBITS to Trilinos, then when you configure Trilinos with:

```
./do-configure -DTrilinos_TRIBITS_DIR:STRING=TriBITS/tribits \
  -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES=OFF -DTrilinos_ENABLE_MiniTensor=ON
```

you will see the configure output containing:

```

Enabling all optional intra-package enables <TRIBITS_PACKAGE>_ENABLE_<DEPPACKAGE> that are not currently disabled if both sets of packages are enabled ...

...
-- NOT setting Sacado_ENABLE_Teuchos=ON since Teuchos is NOT enabled at this point!
...

...

Enabling the shell of non-enabled parent packages (mostly for show) that have at least one subpackage enabled ...

-- Setting Trilinos_ENABLE_Kokkos=ON because Trilinos_ENABLE_KokkosCore=ON
-- Setting Trilinos_ENABLE_Teuchos=ON because Trilinos_ENABLE_TeuchosCore=ON
```

**Comment by bartlettroscoe** [comment]
I made several more commits that dependency handling logic printouts related to this issue.  Hopefully the output will be a little clearer than it was before this issue was created.

**Comment by bartlettroscoe** [comment]
I introduced a defect in one of these commits that broke the build of Panzer on a bunch of platforms due to EpetraExt_ENABLE_HDF5=OFF **NOT** turning off.  It was a simple copy and paste error that I fixed in the TriBITS commit 3fe4ff1 and snapshotted to Trillinos in the Trilinos commit https://github.com/trilinos/Trilinos/commit/12b2cdd08767ae26666f46a817c9b0fa6a1c4b5e.

The current automated tests do not adequately catch improper setting of the variables `<Package>_ENABLE_<DepPackage>` or `<Package>_ENABLE_<TPL>`.  They only check for the print out of the logic (which is not the same thing).  However, the current tests for the dependency logic for setting `<Project>_ENABLE_<Package>` and `<Project>_ENABLE_<TPL>` are strong because they grep the later `Final set of [enabled|disabled] [packages|TPLs]` lines that show what get enabled or not enabled.  There is no such independent printout for variables like `<Package>_ENABLE_<DepPackage>` or `<Package>_ENABLE_<TPL>`.  I am not sure the best way to write automated tests for these variables but if something like this comes up again, I will.  But the set statements are literally on the line right below the print statements for the logic.

**Comment by bartlettroscoe** [comment]
I think the problems with this are now resolved.  Therefore, I think it is safe to take this out of review and close this.
